File: assignment34/hw3/getNamePosition.py
# 读取文件并建立姓名到记录起始位置的映射关系
name_pointers = {}

import sys
import json

if len(sys.argv) != 2:
    sys.exit(1)

file_path = sys.argv[1]

with open(file_path, 'r') as file:
    current_pos = 0
    line = file.readline()

    while line:
        # 记录当前姓名对应的起始位置
        name = list(eval(line).keys())[0]
        name_pointers[name] = current_pos

        # 更新当前位置，准备读取下一行
        current_pos = file.tell()
        line = file.readline()

# 打印姓名到记录起始位置的映射关系
for name, pointer in name_pointers.items():
    print(f'{name}: {pointer}')

# 验证
given_name = 'mvv'

if given_name in name_pointers:
    with open(file_path, 'r') as file:
        file.seek(name_pointers[given_name])
        json_data = json.loads(file.readline())
        print(json.dumps(json_data, indent=2))
else:
    print("no this name")

File: assignment34/hw3/src/main/java/cn/edu/tongji/tools/GetNamePosition.java
package cn.edu.tongji.tools;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class GetNamePosition {
    public static Map<String, Long> getNamePointer(String filePath) {
        Map<String, Long> namePointers = new HashMap<>();

        try (RandomAccessFile file = new RandomAccessFile(filePath, "rw")) {
            long currentPos = 0;
            String line = file.readLine();

            while (line != null) {
                // 记录当前姓名对应的起始位置
                String name = line.substring(2, line.indexOf("\":"));
                namePointers.put(name, currentPos);

                // 更新当前位置，准备读取下一行
                currentPos = file.getFilePointer();
                line = file.readLine();
            }
        } catch (IOException e) {
            e.printStackTrace();
            System.exit(1);
        }

        // 打印姓名到记录起始位置的映射关系
//        for (Map.Entry<String, Long> entry : namePointers.entrySet()) {
//            System.out.println(entry.getKey() + ": " + entry.getValue());
//        }
        return namePointers;
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String filePath = sc.nextLine();

        var namePointers = getNamePointer(filePath);

        // 验证
        String givenName = "mvv";
        if (namePointers.containsKey(givenName)) {
            try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
                reader.skip(namePointers.get(givenName));
                String jsonData = reader.readLine();
                System.out.println(jsonData);
            } catch (IOException e) {
                e.printStackTrace();
            }
        } else {
            System.out.println("no this name");
        }
    }
}
